Adds exception handling to the replace_secrets.rb and provides a helpful resolution message

emilylaguna opened the pull request:
This adds some additional messaging when `replace_secrets.rb` fails to generate, and provides a resolution message.

This also disables showing all the environment variables when the script is ran.

## Screenshots

### Before
<img width="1322" alt="Screen Shot 2023-02-06 at 10 51 29 AM" src="https://user-images.githubusercontent.com/793774/217018911-9dfc963f-767a-47fa-8cbb-9ad9512fe676.png">

### After
<img width="1253" alt="Screen Shot 2023-02-06 at 10 50 40 AM" src="https://user-images.githubusercontent.com/793774/217018726-8fa4681b-3910-4bce-b8cf-effb3d8b8e87.png">


## To test

1. In Xcode open `ApiCredentials.tpl`
2. Add a line like: `static let test = "%{test_key}"` to the bottom
3. Open ApiCredentials.swift
4. Right Click it and Show it in Finder
5. Delete it
6. Run the build again, it should fail
7. View the build details and locate the PreBuildActions step
8. Click the ☰ icon to show the error details
9. ✅ Verify you see the more detailed failure message

## Checklist

- [x] I have considered if this change warrants user-facing release notes and have added them to `CHANGELOG.md` if necessary.
- [x] I have considered adding unit tests for my changes.
- [x] I have updated (or requested that someone edit) [the spreadsheet](https://docs.google.com/spreadsheets/d/107jqrutZhU0fVZJ19SBqxxVKbV2NWSdQC9MFYdLiAxc/edit?usp=sharing) to reflect any new or changed analytics.

emilylaguna committed:
Adds exception handling to the replace_secrets.rb and provides a helpful resolution message

emilylaguna committed:
Change error messaging

emilylaguna commented on podcasts/Credentials/replace_secrets.rb:
```diff
@@ -53,14 +53,21 @@ def load(secrets_path)
+    end
+    template.close
+  rescue => exception
+    STDERR.puts("\n🚨🚨 Failed to generate credentials file from template: " + File.basename(template_path) +  " 🚨🚨")
```
Everyone knows emoji's are necessary. 🦄

pocketcasts replied:
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/Automattic/pocket-casts-ios/blob/8f9153702e19dd5069e7d905af01d0f543fc0bb3/podcasts/Credentials/replace_secrets.rb#L65"></span>Operator <code>+</code> should be surrounded by a single space.</td>
    </tr>
  </tbody>
</table>

pocketcasts replied:
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/Automattic/pocket-casts-ios/blob/8f9153702e19dd5069e7d905af01d0f543fc0bb3/podcasts/Credentials/replace_secrets.rb#L65"></span>Unnecessary spacing detected.</td>
    </tr>
  </tbody>
</table>

pocketcasts replied:
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/Automattic/pocket-casts-ios/blob/8f9153702e19dd5069e7d905af01d0f543fc0bb3/podcasts/Credentials/replace_secrets.rb#L65"></span>Prefer single-quoted strings when you don't need string interpolation or special symbols.</td>
    </tr>
  </tbody>
</table>

pocketcasts replied:
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/Automattic/pocket-casts-ios/blob/8f9153702e19dd5069e7d905af01d0f543fc0bb3/podcasts/Credentials/replace_secrets.rb#L65"></span>Use <code>$stderr</code> instead of <code>STDERR</code>.</td>
    </tr>
  </tbody>
</table>

pocketcasts replied:
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/Automattic/pocket-casts-ios/blob/8f9153702e19dd5069e7d905af01d0f543fc0bb3/podcasts/Credentials/replace_secrets.rb#L65"></span>Use <code>warn</code> instead of <code>STDERR.puts</code> to allow such output to be disabled.</td>
    </tr>
  </tbody>
</table>

pocketcasts replied:
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/Automattic/pocket-casts-ios/blob/8f9153702e19dd5069e7d905af01d0f543fc0bb3/podcasts/Credentials/replace_secrets.rb#L65"></span>Prefer string interpolation to string concatenation.</td>
    </tr>
  </tbody>
</table>

leandroalonso commented:
cc @mokagio

pocketcasts commented on podcasts/Credentials/replace_secrets.rb:
```diff
@@ -53,14 +53,21 @@ def load(secrets_path)
 def process(template_path, secrets_path)
-  secrets = load(secrets_path)
-  template = File.open(template_path, 'r')
+  begin
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/Automattic/pocket-casts-ios/blob/8f9153702e19dd5069e7d905af01d0f543fc0bb3/podcasts/Credentials/replace_secrets.rb#L56"></span>Redundant <code>begin</code> block detected.</td>
    </tr>
  </tbody>
</table>

pocketcasts commented on podcasts/Credentials/replace_secrets.rb:
```diff
@@ -53,14 +53,21 @@ def load(secrets_path)
+      puts line % secrets
+    end
+    template.close
+  rescue => exception
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/Automattic/pocket-casts-ios/blob/8f9153702e19dd5069e7d905af01d0f543fc0bb3/podcasts/Credentials/replace_secrets.rb#L64"></span>Use <code>e</code> instead of <code>exception</code>.</td>
    </tr>
  </tbody>
</table>

pocketcasts replied:
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/Automattic/pocket-casts-ios/blob/8f9153702e19dd5069e7d905af01d0f543fc0bb3/podcasts/Credentials/replace_secrets.rb#L64"></span>Avoid rescuing without specifying an error class.</td>
    </tr>
  </tbody>
</table>

pocketcasts commented on podcasts/Credentials/replace_secrets.rb:
```diff
@@ -53,14 +53,21 @@ def load(secrets_path)
+    template.close
+  rescue => exception
+    STDERR.puts("\n🚨🚨 Failed to generate credentials file from template: " + File.basename(template_path) +  " 🚨🚨")
+    STDERR.puts("\n-> Exception: " + exception.message)
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/Automattic/pocket-casts-ios/blob/8f9153702e19dd5069e7d905af01d0f543fc0bb3/podcasts/Credentials/replace_secrets.rb#L66"></span>Use <code>warn</code> instead of <code>STDERR.puts</code> to allow such output to be disabled.</td>
    </tr>
  </tbody>
</table>

pocketcasts replied:
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/Automattic/pocket-casts-ios/blob/8f9153702e19dd5069e7d905af01d0f543fc0bb3/podcasts/Credentials/replace_secrets.rb#L66"></span>Use <code>$stderr</code> instead of <code>STDERR</code>.</td>
    </tr>
  </tbody>
</table>

pocketcasts replied:
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/Automattic/pocket-casts-ios/blob/8f9153702e19dd5069e7d905af01d0f543fc0bb3/podcasts/Credentials/replace_secrets.rb#L66"></span>Prefer string interpolation to string concatenation.</td>
    </tr>
  </tbody>
</table>

pocketcasts commented on podcasts/Credentials/replace_secrets.rb:
```diff
@@ -53,14 +53,21 @@ def load(secrets_path)
+  rescue => exception
+    STDERR.puts("\n🚨🚨 Failed to generate credentials file from template: " + File.basename(template_path) +  " 🚨🚨")
+    STDERR.puts("\n-> Exception: " + exception.message)
+    STDERR.puts("-> Reason: Secrets are most likely out of date.")
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/Automattic/pocket-casts-ios/blob/8f9153702e19dd5069e7d905af01d0f543fc0bb3/podcasts/Credentials/replace_secrets.rb#L67"></span>Use <code>warn</code> instead of <code>STDERR.puts</code> to allow such output to be disabled.</td>
    </tr>
  </tbody>
</table>

pocketcasts replied:
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/Automattic/pocket-casts-ios/blob/8f9153702e19dd5069e7d905af01d0f543fc0bb3/podcasts/Credentials/replace_secrets.rb#L67"></span>Use <code>$stderr</code> instead of <code>STDERR</code>.</td>
    </tr>
  </tbody>
</table>

pocketcasts replied:
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/Automattic/pocket-casts-ios/blob/8f9153702e19dd5069e7d905af01d0f543fc0bb3/podcasts/Credentials/replace_secrets.rb#L67"></span>Prefer single-quoted strings when you don't need string interpolation or special symbols.</td>
    </tr>
  </tbody>
</table>

pocketcasts commented on podcasts/Credentials/replace_secrets.rb:
```diff
@@ -53,14 +53,21 @@ def load(secrets_path)
+    STDERR.puts("\n🚨🚨 Failed to generate credentials file from template: " + File.basename(template_path) +  " 🚨🚨")
+    STDERR.puts("\n-> Exception: " + exception.message)
+    STDERR.puts("-> Reason: Secrets are most likely out of date.")
+    STDERR.puts("-> Solution: Run: bundle exec fastlane run configure_apply\n\n")
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/Automattic/pocket-casts-ios/blob/8f9153702e19dd5069e7d905af01d0f543fc0bb3/podcasts/Credentials/replace_secrets.rb#L68"></span>Use <code>$stderr</code> instead of <code>STDERR</code>.</td>
    </tr>
  </tbody>
</table>

pocketcasts replied:
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/Automattic/pocket-casts-ios/blob/8f9153702e19dd5069e7d905af01d0f543fc0bb3/podcasts/Credentials/replace_secrets.rb#L68"></span>Use <code>warn</code> instead of <code>STDERR.puts</code> to allow such output to be disabled.</td>
    </tr>
  </tbody>
</table>

mokagio committed:
Run RuboCop with autocorrect

This addresses violations picked up by Danger in
https://github.com/Automattic/pocket-casts-ios/pull/693.

mokagio committed:
Address RuboCop method length warning in `replace_secrets.rb`

This is something the autocorrect in the previous commit couldn't
address.

See 6a94f75205672c7b5525b58ddf83c04ed0a14ce6

mokagio reviewed (approved):
Brilliant. Thanks @emilylaguna

leandroalonso committed:
Merge pull request #696 from Automattic/mokagio/address-rubocop-693

Run RuboCop with autocorrect on top of #693
